### ISSUE TITLE
Not working on Android 14?

### ISSUE THREAD

**Issue by eirinn1975** [rdoeffinger/Dictionary#159]
The app stopped working after updating to Android 14. The splash screen appears for a fraction of second and then it closes. Is it just me?

**Comment by rdoeffinger** [comment]
I do not have an actual device running Android 14.
But it seems to work fine on the emulator.
I also see no crash reports that seem related on the Play Store side either.
I am working on an update to fully support Android 14. Those are just minor detail changes though, but maybe it would fix something.
In the mean time, the usual cleaning cache and data and/or reinstalling might be worth a try?

**Comment by eirinn1975** [comment]
I've tried cleaning cache, purging the storage and reinstalling, but nothing worked.

**Comment by rdoeffinger** [comment]
Are you able to get access to the logical, or some other way to get a backtrace?
Or if you have the Play Store version to allow Google to send error reports?
I have one single crash report from Android 14 on a Pixel 7 Pro, but that seems to be something broken in Android's libicu and not the app itself...

**Comment by eirinn1975** [comment]
Interesting, I'm on a pixel 7 (not pro). I removed the store version and installed it manually, but I will go back to the store version and allow error reports. Any fix for that libicu thingy?



--

Maximiliano Morsia

Hakamäentie 31, 26840 Rauma - Finland

tel. +358 (0) 40 175 1973<tel:+358401751973>

________________________________

**Comment by rdoeffinger** [comment]
You can also opt in to the 5.7.0 testing release to see if that helps.
I think the libicu thing was some OS level bug or data corruption, it was on the beta of Android 14, not final version.
So I think it's unlikely to be related to your issue, and I certainly have no idea what could be done about it (and not planning to try).

**Comment by eirinn1975** [comment]
Reinstalled the app from the store. Both stable (5.6.3) and beta version (5.7.0) crash on startup. I've forwarded the collected data after crashes. Hopefully it will be useful.

**Comment by rdoeffinger** [comment]
Thanks for testing! Have not gotten the crash data yet, will check again tomorrow.

**Comment by eirinn1975** [comment]
Just for info, I don't if it can help, but the exception class name seems to be java.lang.StackOverflowError: 
 
![Screenshot_20231012-183808](https://github.com/rdoeffinger/Dictionary/assets/11836782/80f2f1a4-ef0c-43dc-9b75-6ab56361d839)

**Comment by rdoeffinger** [comment]
Ah, useful, so it is indeed that ICU issue.
If lucky, it's some kind of startup race condition, I am adding a workaround, see patch linked above.

**Comment by rdoeffinger** [comment]
Published a version with that workaround as beta (5.7.1). Please report if it changes anything

**Comment by eirinn1975** [comment]
Thanks! I'll let you know as soon as it hits the f-droid or Google store.

**Comment by eirinn1975** [comment]
Yes! It's working now 😊 thanks a lot!

**Comment by eirinn1975** [comment]
Oh not so quick... After downloading a dictionary and selecting it, it went back to the same problem.

**Comment by rdoeffinger** [comment]
Yes, that will be when ICU is initialised when the other code fails.
Ok, that means it's indeed an ICU issue and it's not a race condition.
Unfortunately I am out of ideas for now.

**Comment by rdoeffinger** [comment]
Do you have the ability to experiment with your locale (e.g. app language) setting?
I just wonder if this maybe only happens in certain regions, and if that could be why I cannot reproduce it.

**Comment by eirinn1975** [comment]
I could change it, yes, however my phone system language is set to English (United States). Any idea which language should I try?

**Comment by rdoeffinger** [comment]
Ok, that I should have tested, so it's probably not it. I guess you could try English (UK) or such, but maybe not worth messing with your phone settings since it's unlikely to help.
I tried having a look at the ICU code to see if I could limit the recursion somehow, but did not find any there either so far.

**Comment by eirinn1975** [comment]
I've given UK English, Italian and Finnish a try, but all with the same result.

**Comment by RDusr05** [comment]
I have exactly the problem with Android 14

**Comment by occivink** [comment]
As the only competent FOSS offline dictionary app, it's a shame that it won't run on android 14 at all. I can't contribute on the android development side, but I'd be happy to donate money for this bug to be fixed.

**Comment by rdoeffinger** [comment]
What you can do to help as a non-developer:
- test non-Google Android 14 device. All the issue reports I have so far are on Google devices
- try not convince Google to investigate this. Not very likely to work, but if there's evidence that non-Google Android 14 devices work fine, that might increase the chances.

I have no conclusive proof, but my best guess is that this is a bug in Android.
If I had access to an affected device, I might be able to come up with a workaround, but since the app would still not work 100% right, it's not worth a lot of effort.

**Comment by rdoeffinger** [comment]
I now have confirmation that the app works fine on a Samsung S22 running Android 14 (and previously confirmed that it works fine on emulator with Android 14), making it more likely that this is a Google issue.
At least it is not a general Android 14 incompatibility.

**Comment by bypear** [comment]
Another affected user here. Using a Pixel 6, the issues happens with the stock OS and also with a 3rd party OS. There's no other translator app I would want to use. So, please don't let use hang with this issue. Pretty please. ;)

Besides, how can it be that it works on Samsung but not on pixels? If it really is an Android 14 specific bug, then I wonder that Samsung patched that bug out. My guess, somehow luckily it works with Samsung but probably more devices will come where it fails.

**Comment by rdoeffinger** [comment]
> So, please don't let use hang with this issue. Pretty please. ;)
It's not so simple. Right now I'm enjoying a mostly computer-free vacation.
But after that, either I can blindly try out workarounds and hope someone among you can test them for me.
Or I find someone with an affected device so I can experiment hands-on, which would probably speed it up a lot.
That option is a lot harder if only pixel devices are affected, thus my interest in reports of other devices with issues.

> Besides, how can it be that it works on Samsung but not on pixels?
The Android emulator by Google also works fine.

> then I wonder that Samsung patched that bug out.
More likely, the Google pixel team patched it in...

> somehow luckily it works with Samsung but probably more devices will come where it fails.
That is also possible. And that would be useful information.

**Comment by atopia** [comment]
I can confirm the issue (downloading and unpacking dictionaries works, but the app crashes as soon as a dictionary is selected) on a Fairphone 4 with CalyxOS 5.3.1, based on AOSP 14. If the Let me know if there is anything I can do to help debug the issue. The app worked great (thank you!)  on Android 13 based CalyxOS and stopped working right with the update to Android 14. I would guess that the similarity to the affected Pixels is in the shared AOSP code base that somehow does not apply to Samsung's Android variant, but I unfortunately I don't have deeper knowledge of Android.

**Comment by rdoeffinger** [comment]
This is a pretty bad hack, but maybe someone can test if this version works around the issue?
I expect it will create some issues with word search, but it's a try
[Dictionary-issue159-testing.zip](https://github.com/rdoeffinger/Dictionary/files/13969122/Dictionary-issue159-testing.zip)

**Comment by eirinn1975** [comment]
I've just tried the testing version. Unfortunately it doesn't solve the issue. If anything, now the app crashes whimsically at startup or after selecting the dictionary.

**Comment by atopia** [comment]
> This is a pretty bad hack, but maybe someone can test if this version works around the issue?

I can confirm that this version still exhibits the same crashing behavior as the latest beta.

**Comment by code-consensus** [comment]
Hi, I have another minor detail that may also provide some insight into why the problem exists.

In my [prior post above](#issuecomment-1914437073), I mentioned how in some user profiles _on the same phone_, the problem didn't exist. One thing I noticed in the "dysfunctional" user profile, was that after installation and program launch, there is no data directory "accessible" somehow.

Steps to reproduce:
1. install Quickdic
2. launch the program and adjust settings
3. open the Files app, and go to  /storage/emulated/{userID}/Android/data and no directory of "de.reimardoeffinger.quickdic" is shown to exist (though doing the same install on another user profile _does_ show this folder created)
4. launch app again and download a dictionary -- the app will show messages that a dictionary gets downloaded and installed, and then it will appear in the "dictionaries on device" list
5. again, using the Files app, "de.reimardoeffinger.quickdic" somehow doesn't show up
6. connect the phone to a computer for USB file transfer -- browsing to \Internal shared storage\Android\data and no "de.reimardoeffinger.quickdic" directory is shown

The app has clearly downloaded dictionary files and shows them being stored. So they must be there somehow, but maybe the fact that it is not visible is pointing to the problem.

Hopefully that gives some hint as to what is happening

**Comment by G-i-o** [comment]
Same issue here.  Worked fine until I upgraded to Lineage 21 (Android 14 based).

It crashes only after downloading the dictionaries. Changing the data folder has not solved it either.

Tested device: Google Pixel 3a.

Thank You.

**Comment by basings** [comment]
Same behavior as G-i-o on FrapheneOS A14, fdroid version

**Comment by code-consensus** [comment]
An update to my comment above, regarding the app working on certain profiles versus others -- **this is no longer the case**. The app no longer works at all in _any_ profile for me, even with a brand new profile with nothing else installed.

I'm not sure what mitigating factor was involved before, and what changed with Android since [my post](https://github.com/rdoeffinger/Dictionary/issues/159#issuecomment-1914437073), but this is the case as of at least the latest Pixel version (AP1A.240505.004) and perhaps an earlier update too.

**Comment by altercat10n** [comment]
This problem persists for almost a year. How do we resolve it?

(Is there another FLOSS offline dictionary app, preferably in F-Droid?)

**Comment by rdoeffinger** [comment]
Sorry, I know it's frustrating and it does not help you, but the app works 100% perfectly on all phones I have access to (I even asked friends to try), as well the Android emulator.
I did have one more idea to try though, is someone wants to try the attached version if it changes anything...
[Dictionary-issue159-test2.zip](https://github.com/user-attachments/files/16680910/Dictionary-issue159-test2.zip)

**Comment by eirinn1975** [comment]
> Sorry, I know it's frustrating and it does not help you, but the app works 100% perfectly on all phones I have access to (I even asked friends to try), as well the Android emulator. I did have one more idea to try though, is someone wants to try the attached version if it changes anything... [Dictionary-issue159-test2.zip](https://github.com/user-attachments/files/16680910/Dictionary-issue159-test2.zip)

I've tried to install this version, but my pixel with android 14 returns "you can't install the app on your device"

**Comment by G-i-o** [comment]
> > Sorry, I know it's frustrating and it does not help you, but the app works 100% perfectly on all phones I have access to (I even asked friends to try), as well the Android emulator. I did have one more idea to try though, is someone wants to try the attached version if it changes anything... [Dictionary-issue159-test2.zip](https://github.com/user-attachments/files/16680910/Dictionary-issue159-test2.zip)
> 
> I've tried to install this version, but my pixel with android 14 returns "you can't install the app on your device"

You probably need to change permissions in the settings.

By default, Android does not let you install applicazions from sources other than Google Play.

**Comment by G-i-o** [comment]
I will try this as soons as I can. Thank You for working on it @rdoeffinger , whether it works or not. Will report.

Meanwhile, @altercat10n , there's another dictionary on F-Droid. It's called Aard2. It's only a simple dictionary, though (eg. no translator).

**Comment by eirinn1975** [comment]
> > > Sorry, I know it's frustrating and it does not help you, but the app works 100% perfectly on all phones I have access to (I even asked friends to try), as well the Android emulator. I did have one more idea to try though, is someone wants to try the attached version if it changes anything... [Dictionary-issue159-test2.zip](https://github.com/user-attachments/files/16680910/Dictionary-issue159-test2.zip)
> > 
> > 
> > I've tried to install this version, but my pixel with android 14 returns "you can't install the app on your device"
> 
> You probably need to change permissions in the settings.
> 
> By default, Android does not let you install applicazions from sources other than Google Play.

Must be something else. I can install other apks just fine, but not this one.

**Comment by G-i-o** [comment]
> Sorry, I know it's frustrating and it does not help you, but the app works 100% perfectly on all phones I have access to (I even asked friends to try), as well the Android emulator. I did have one more idea to try though, is someone wants to try the attached version if it changes anything... [Dictionary-issue159-test2.zip](https://github.com/user-attachments/files/16680910/Dictionary-issue159-test2.zip)

I can confirm that this version works.

Thank You so much @rdoeffinger for your work! :pray:

**Comment by G-i-o** [comment]
@eirinn1975 , have you uninstalled the previous version? That is the other thing it may be happening as they cannot be installed together and they have a different signature, thus, your package manger won't update the other with this.

**Comment by atopia** [comment]
> Sorry, I know it's frustrating and it does not help you, but the app works 100% perfectly on all phones I have access to (I even asked friends to try), as well the Android emulator. I did have one more idea to try though, is someone wants to try the attached version if it changes anything... [Dictionary-issue159-test2.zip](https://github.com/user-attachments/files/16680910/Dictionary-issue159-test2.zip)

Works for me too. Amazing work, @rdoeffinger!

**Comment by altercat10n** [comment]
> I did have one more idea to try though, is someone wants to try the attached version if it changes anything... [Dictionary-issue159-test2.zip](https://github.com/user-attachments/files/16680910/Dictionary-issue159-test2.zip)

Installed the app, downloaded dictionaries, searched for some words. This seems to work (GrapheneOS Stable (Android 14), Build 2024080600). Thank you!

**Comment by BrilliantTrees** [comment]
I was directed here via my comment in #166 

After installing the new APK provided above all appears to be working now. 

Thanks @rdoeffinger

**Comment by somogyic** [comment]
> Sorry, I know it's frustrating and it does not help you, but the app works 100% perfectly on all phones I have access to (I even asked friends to try), as well the Android emulator. I did have one more idea to try though, is someone wants to try the attached version if it changes anything... [Dictionary-issue159-test2.zip](https://github.com/user-attachments/files/16680910/Dictionary-issue159-test2.zip)

wfm too, thanks @rdoeffinger

**Comment by code-consensus** [comment]
This solves the problem for me too!

The only thing I did notice with this test APK is that the app had some responsiveness issue -- though this may be because the file is not a release variant. 

Specifically, if you press the flag icon at the top left to select a different dictionary, it takes about 5 seconds for the dictionary list to show up (on my phone -- Pixel 6 with Android v14) . But once you are presented with the list of dictionaries, pressing the flag of the dictionary you would like to switch to brings up the dictionary instantly.

In any case, this is a small price to pay for a functional app!

@rdoeffinger out of pure curiosity, given the app was working for the vast majority of phones, how did you find the problem and what did you change in the end?

> Sorry, I know it's frustrating and it does not help you, but the app works 100% perfectly on all phones I have access to (I even asked friends to try), as well the Android emulator. I did have one more idea to try though, is someone wants to try the attached version if it changes anything... [Dictionary-issue159-test2.zip](https://github.com/user-attachments/files/16680910/Dictionary-issue159-test2.zip)

**Comment by nutpantz** [comment]
Pixel 8 pro Android 14. 
This does not work for me. 
There is no response at all, no error nothing

**Comment by rdoeffinger** [comment]
Short version: unfortunately this is a good hint, but not enough to do an acceptable fix, so I need you to do some more testing for me.
This one just updates the ICU version but still reduces its size. If it works, we're done. If it does not work, then it gets tricky and a lot more work... Please test.
[Dictionary-issue159-test3.zip](https://github.com/user-attachments/files/16801562/Dictionary-issue159-test3.zip)


More details:
Sorry for no response, I was travelling right after uploading this, and github's advanced security requirements for devs meant I was not able to respond.
I thought the crashingissue was with a system component, but I had completely forgotten that some time ago I replaced it with a custom copy.
What the test version does is update the version and package the full version instead of the stripped down version of ICU.
The latter part increases the app size a lot and probably causes performance issues (@code-consensus it is a release variant) .
As to the installation issue, this uses my signing key, which is compatible with app store releases, but F-Droid releases use their own key, that's likely why some of you had to uninstall first. Or it could be because I forgot to increase the version number...

**Comment by nutpantz** [comment]
Here is a log,i don't see it acts any different

1724946729.150 10329 29725 29725 D VRI[DictionaryManagerActivity]: visibilityChanged oldVisibility=false newVisibility=true
1724946729.181 10329 29725 29725 D AutofillManager: notifyViewReadyInner:1073741825
1724946729.214 10329 29725 29725 I ImeTracker: de.reimardoeffinger.quickdic:ae5f45fe: onRequestHide at ORIGIN_CLIENT_HIDE_SOFT_INPUT reason HIDE_SOFT_INPUT_FROM_VIEW
1724946729.215 10329 29725 29725 I ImeTracker: de.reimardoeffinger.quickdic:ae5f45fe: onFailed at PHASE_CLIENT_VIEW_SERVED
1724946729.215 10329 29725 29725 W InputMethodManager: Ignoring hideSoftInputFromView() as view=androidx.appcompat.widget.f{1b26963 VFED..C.. ......I. 0,0-0,0} is not served.
1724946729.218 10329 29725 29725 D VRI[DictionaryManagerActivity]: Start draw after previous draw not visible
1724946729.277 10329 29725 29725 I ImeTracker: de.reimardoeffinger.quickdic:da1c71c6: onRequestHide at ORIGIN_CLIENT_HIDE_SOFT_INPUT reason HIDE_SOFT_INPUT_FROM_VIEW
1724946729.277 10329 29725 29725 I ImeTracker: de.reimardoeffinger.quickdic:da1c71c6: onFailed at PHASE_CLIENT_VIEW_SERVED
1724946729.277 10329 29725 29725 W InputMethodManager: Ignoring hideSoftInputFromView() as view=androidx.appcompat.widget.f{de1f209 VFED..C.. ......I. 0,0-0,0} is not served.
1724946732.385 10329 29725 29725 D QuickDic: Downloading to: ES.quickdic.v007.zip
1724946732.516 10329 29725 29725 W QuickDic: Download started: ES.quickdic.v007.zip
1724946733.579 10329 29725 29725 D QuickDic: Downloading to: EN-AF.quickdic.v007.zip
1724946733.706 10329 29725 29725 W QuickDic: Download started: EN-AF.quickdic.v007.zip
1724946735.350 10329 29725 29725 D QuickDic: Downloading to: EN-grc.quickdic.v007.zip
1724946735.476 10329 29725 29725 W QuickDic: Download started: EN-grc.quickdic.v007.zip
1724946739.307 10329 29725 29725 D VRI[DictionaryManagerActivity]: visibilityChanged oldVisibility=true newVisibility=false
1724946739.328 10329 29725 29725 D VRI[DictionaryManagerActivity]: Not drawing due to not visible

**Comment by rdoeffinger** [comment]
@nutpantz I think your problem is a different one from the stack crash the others experienced (the other reports were crashes when opening a dictionary, not hangs or download list issues). I have no idea what the cause might be and if there is a way to avoid/fix it.

**Comment by nutpantz** [comment]
It was crashing, but i cleared storage and it stopped crashing

**Comment by code-consensus** [comment]
Hi @rdoeffinger , I tested your test3 version as requested.

This version _also_ **works**. It doesn't crash on open, and it doesn't crash on loading a dictionary.

_Please note_, the performance issue I noted before is still there -- clicking the flag at the top left to change the dictionary takes several seconds to load the list of dictionaries.

**However**, I will also note that if instead of clicking the flag to switch dictionaries, you click the three dots at the top right and then "Dictionary manager", the list loads near instantly.

> Short version: unfortunately this is a good hint, but not enough to do an acceptable fix, so I need you to do some more testing for me. This one just updates the ICU version but still reduces its size. If it works, we're done. If it does not work, then it gets tricky and a lot more work... Please test. [Dictionary-issue159-test3.zip](https://github.com/user-attachments/files/16801562/Dictionary-issue159-test3.zip)
> 
> More details: Sorry for no response, I was travelling right after uploading this, and github's advanced security requirements for devs meant I was not able to respond. I thought the crashingissue was with a system component, but I had completely forgotten that some time ago I replaced it with a custom copy. What the test version does is update the version and package the full version instead of the stripped down version of ICU. The latter part increases the app size a lot and probably causes performance issues (@code-consensus it is a release variant) . As to the installation issue, this uses my signing key, which is compatible with app store releases, but F-Droid releases use their own key, that's likely why some of you had to uninstall first. Or it could be because I forgot to increase the version number...

**Comment by G-i-o** [comment]
> @nutpantz I think your problem is a different one from the stack crash the others experienced (the other reports were crashes when opening a dictionary, not hangs or download list issues). I have no idea what the cause might be and if there is a way to avoid/fix it.

Yes, at least from my side, I can report that I've never had any issues about app hanging and other issues except the one reported here (app crashing on dictionary load), which was apparently fixed with the test version posted here.

**Comment by rdoeffinger** [comment]
This version has been released to Google Play in the testing (beta) track.
I assume F-Droid will soon build and make it available as well (5.7.2 is the version).
I will give it a bit of time, check that it seems to fix things, release to main track on Google play.
Assuming no issues, I expect to close this ticket soon.
Issues that are not a crash to due a stack overflow issue will need their own ticket.
I am not sure the slow showing of dictionary list is a new issue or not, on my phone it's plenty fast, but on the emulator it is a bit slow. If I ever have some time over I might investigate a bit.

**Comment by rdoeffinger** [comment]
Btw in case you are also curious: there are around 4000 users that have the app installed on a device with Android 14.
But there are only about 6 crashes per month in the time since this issue appeared, so less than 100 crashes.
While this is the most serious issue the app had, with "only" 5% of users affected it's not trivial to investigate, and it would not have been possible without you helping me to test things!

**Comment by rdoeffinger** [comment]
Unfortunately I am still seeing the same crash with the newly released version for at least some users. Any ideas from anyone?

**Comment by thranur** [comment]
Thank you very much for trying to fix this issue. I have been using Quickdic on my phones for years, and just got a Pixel 6 and it is crashing. I just tried the 5.7.2 APK from GitHub and it's still crashing after downloading a dictionary. Clearing cache does not help, I have to clear storage and then I can try to download a different dictionary. I tried a few but no changes. I can try using ADB to debug or something but I have never developed on Android.

**Comment by thranur** [comment]
Sorry if this is useless to you, but just in case, I managed to compile the latest git (v5.7.3) with Android Studio and run it in debug mode, and after downloading a dictionary and trying to open it, I still have the crash, and besides the stack overflow with no useful information (`at n0.p.b0(SourceFile:224)`), I also have:

```
2024-09-07 21:53:46.323 28154-28154 QuickDic                de.reimardoeffinger.quickdic.debug   D  onSearchTextChange: 
2024-09-07 21:53:46.324 28154-28154 QuickDic                de.reimardoeffinger.quickdic.debug   D  Hide soft keyboard.
2024-09-07 21:53:46.324 28154-28214 .quickdic.debug         de.reimardoeffinger.quickdic.debug   I  Rejecting re-init on previously-failed class java.lang.Class<com.ibm.icu.text.Transliterator>: java.lang.StackOverflowError: stack size 1038KB
                                                                                                    (Throwable with no stack trace)
2024-09-07 21:53:46.325 28154-28154 ImeTracker              de.reimardoeffinger.quickdic.debug   I  de.reimardoeffinger.quickdic.debug:44ef91cf: onRequestHide at ORIGIN_CLIENT_HIDE_SOFT_INPUT reason HIDE_SOFT_INPUT
2024-09-07 21:53:46.326 28154-28154 ImeTracker              de.reimardoeffinger.quickdic.debug   I  de.reimardoeffinger.quickdic.debug:44ef91cf: onFailed at PHASE_CLIENT_VIEW_SERVED
2024-09-07 21:53:46.326 28154-28154 ImeTracker              de.reimardoeffinger.quickdic.debug   I  de.reimardoeffinger.quickdic.debug:b2d6eb66: onRequestHide at ORIGIN_CLIENT_HIDE_SOFT_INPUT reason HIDE_SOFT_INPUT
2024-09-07 21:53:46.327 28154-28154 ImeTracker              de.reimardoeffinger.quickdic.debug   I  de.reimardoeffinger.quickdic.debug:b2d6eb66: onFailed at PHASE_CLIENT_VIEW_SERVED
2024-09-07 21:53:46.327 28154-28154 QuickDic                de.reimardoeffinger.quickdic.debug   D  Trying to restore searchText=
2024-09-07 21:53:46.328 28154-28154 QuickDic                de.reimardoeffinger.quickdic.debug   D  prefs changed: indexShortName
2024-09-07 21:53:46.328 28154-28154 QuickDic                de.reimardoeffinger.quickdic.debug   D  prefs changed: dictFile
FATAL EXCEPTION: Thread-3
  Process: de.reimardoeffinger.quickdic.debug, PID: 28154
  java.lang.NoClassDefFoundError: com.ibm.icu.text.Transliterator
  at com.hughes.android.dictionary.engine.TransliteratorManager.get(TransliteratorManager.java:45)
  at com.hughes.android.dictionary.engine.Index.normalizer(Index.java:108)
  at com.hughes.android.dictionary.engine.Index.getSortComparator(Index.java:119)
  at com.hughes.android.dictionary.engine.Index.findExact(Index.java:289)
  at com.hughes.android.dictionary.DictionaryActivity$5.run(DictionaryActivity.java:547)
  at java.lang.Thread.run(Thread.java:1012)
  Caused by: java.lang.StackOverflowError: stack size 1038KB
```

**Comment by rdoeffinger** [comment]
Before going too deep, is it right that neither test2 nor test3 versions of the app work for you?
More technical details:
It's not quite the right version of the source, but essentially you'd want to look at this code: https://android.googlesource.com/platform/external/icu/+/1537b2f/android_icu4j/src/main/java/android/icu/impl/ICUResourceBundle.java
Of particular interest would be the values of the variables in this block (line 1270):
```
                    if(openType == OpenType.LOCALE_DEFAULT_ROOT &&
                            !defaultLocale.getLanguage().equals(localeName)) {
                        b = (ICUResourceBundle)instantiateBundle(baseName, defaultID, root, openType);
```
What I think must be happening is that the language of the default locale never appears in its name, causing endless recursion and stack overflow.
Which is why the 5.7.3 patch tries changing the default locale to US on crash. But seems that does not help.

**Comment by thranur** [comment]
Yes, I tried both 5.7.2 and 5.7.3, and they still crash. Using the debugger and adding a breakpoint at `TransliteratorManager.java`, line 45, `result = Transliterator.createFromRules(...`, I do get to the point in the comment mentioning "// This seems to happen with Android 14 on some Pixel 7 devices", but it seems to happen on Thread-2. Then Thread-3 comes along, I don't really understand what it does, and the application is suddenly terminated and I have the huge stack trace in logcat. So I don't understand if it's the "same" thread that tried again, or something else.

Sorry, I still need to better understand how to use Android Studio. For instance, I am now in UResourceBundle, and it says "Decompiled .class file, bytecode version: 50.0 (Java 6)". I tried clicking "Download" to get the source, but I got "Sources for 'icu4j-57.2.small.jar' not found". I see this line:

    return getBundleInstance(baseName, localeName, ICUResourceBundle.ICU_DATA_CLASS_LOADER, false)

Where `baseName` is `com/ibm/icu/impl/data/icudt57b/translit`, `localeName` is `root`, and `ICUResourceBundle.ICU_DATA_CLASS_LOADER` is `null`.

After advancing some, I got to set a breakpoint at function `instantiateBundle`, at line `b = (ICUResourceBundle)instantiateBundle(baseName, defaultID, root, openType);`. I guess a screenshot will have much more detail than I can get, so here it is:

![Screenshot_20240907_234503](https://github.com/user-attachments/assets/aaffbfda-1899-4345-ab40-378e44746ee7)

So indeed it seems to be recursively calling itself over and over (the number of "hidden frames" increases when the breakpoint is hit).

**Comment by rdoeffinger** [comment]
> Yes, I tried both 5.7.2 and 5.7.3, and they still crash.
I meant the test2 and test3 zip files in this thread

> but I got "Sources for 'icu4j-57.2.small.jar' not found".
The decompiled version seems good enough, but if you want the source you can switch to the full size version instead of small in build.gradle, there is a comment around mksmallicu command.
That also corresponds to the test2.zip I posted, it would be good to know if it happens to fix the issue (though I would prefer not having to include the full size ICU code).

> I guess a screenshot will have much more detail than I can get, so here it is:
Can't really see what exactly defaultLocale is, but localeName = "keyTypeData" is clearly broken, so whatever calculates this goes wrong.

**Comment by thranur** [comment]
Sorry about that, I missed those in the comments.
- test2 works for me;
- test3 crashes.

Concerning the build.gradle command, I saw this:

```
    // Note: you can just replace "smallicu" by implementation
    // and remove the "compile files" part if you do not
    // care about the size of the apk
    smallicu 'com.ibm.icu:icu4j:57.2'
    implementation files("$buildDir/icu4j-57.2.small.jar") {
        builtBy 'makesmallicu'
    }
```

By I didn't understand what "replace smallicu by implementation" means, neither where is the "compile files part".

Anyway, `test2` seems to be working for me (possibly because I have a Pixel 6, not 7), so I guess I can stick with it for now. Thanks again for all the effort. Seeing how alien all of this is to me, I'm even more appreciative of how much effort people put into making and maintaining open-source Android apps.

**Comment by rdoeffinger** [comment]
Ok, based on your information I now published 5.7.4 which includes the full ICU package like test2.
I don't like that it doubles the size of the app without any real benefit besides avoiding this mysterious bug, but if it works I guess it's ok, it's still a fairly small app size.

**Comment by code-consensus** [comment]
> I am not sure the slow showing of dictionary list is a new issue or not, on my phone it's plenty fast, but on the emulator it is a bit slow. If I ever have some time over I might investigate a bit.

Hi @rdoeffinger -- I tested the published v5.7.2 release. At least on my Pixel 6 with Android v14, the app worked _without_ crashing upon loading! 

**_However_**, I do get one _additional_ issue though, which is that I now get an occasional pop-up of "Quickdic not responding" with the option to wait or close the app. The app meanwhile is still working fine if I press "wait", so it is _not_ completely frozen or crashing, though the pop-up will come back frequently after that it seems unless I force close the app and restart. I do wonder if this is maybe related to the slow dictionary loading as I noticed it starting to happen after switching dictionaries.

Here is the the app-not-responding log thread:
`process: de.reimardoeffinger.quickdic`
`activity: de.reimardoeffinger.quickdic/com.hughes.android.dictionary.DictionaryActivity`
`cause: ANR Input dispatching timed out (639973a Select dictionary… (server) is not responding. Waited 5000ms for FocusEvent(hasFocus=true)).`
`info: ANR in de.reimardoeffinger.quickdic (de.reimardoeffinger.quickdic/com.hughes.android.dictionary.DictionaryActivity)`

N.B. I did _not_ download the dictionaries from the internet, but rather had them already downloaded and changed the settings to point the app to the directory where they were stored. I also did not try to download any new dictionaries from the app either.

**Comment by hank-milliken** [comment]
I have GrapheneOS (Android 14) release `2024090400`, and Pixel 7 Pro.


- After upgrade to `5.7.2` the app started to work.
- I started downloading couple of dictionaries
- When I opened the Portuguese dictionaly it app crashed
- Now crashes ever since
- Clearing the cash does not help
- Clearing storage helps

Here is the log I was able to get using: `adb logcat | grep quickdic`

[quicdic-logcat.txt](https://github.com/user-attachments/files/17002161/quicdic-logcat.txt)

**Comment by hank-milliken** [comment]
- After that I downloaded English dictionary and opened it.
- It crashed again

**Comment by rdoeffinger** [comment]
5.7.2 and 5.7.3 are still known broken. As mentioned above you need 5.7.4 for the real fix.

**Comment by code-consensus** [comment]
A bit by accident, I have **_found_ the source of problem**.

I have tested this on v5.7.2 _and_ the prior v5.7.1. I couldn't find the APK for v5.7.4 on github so I couldn't test that version (I checked [releases/tag/v5.7.4](https://github.com/rdoeffinger/Dictionary/releases/tag/v5.7.4) but there was just source-code and I don't have Android Studio).

@rdoeffinger [mentioned above](https://github.com/rdoeffinger/Dictionary/issues/159#issuecomment-2336429465):
> Which is why the 5.7.3 patch tries changing the default locale to US on crash. But seems that does not help.

In the the end, the issue _is_ related to the locale -- **but _NOT_ the language!**

As I mentioned in my posts above, test2, test3 and v5.7.2 all worked for me -- but what I did _not_ take into account was that, [since the last time I had posted in May](https://github.com/rdoeffinger/Dictionary/issues/159#issuecomment-2104125881), there was a fundamental change in my phone: **I had done a factory reset.**

What I discovered was that the test versions and v5.7.2 all worked for me, not because of any changes in those versions, but because of the factory reset. **Because v5.7.1 now _also_ works perfectly for me.**

The problem is _not_ the language -- I have tested it by using US English, German and Spanish -- all worked _perfectly_ **without any crashes**.

**&ast;&ast;&ast; Here is the problem: &ast;&ast;&ast;** 
**The presence of _"Regional preferences"_ causes the app to crash.** _**That**_ is the source of the problem -- if _these_ are not set to "Use default", _then_ the app will crash.

For example, go to System \ Languages and then to Regional preferences, then change Temperature and First day of the week to Celsius and Sunday. This will change your locale from e.g. `en_US` to `en_US_#u-fw-sun-mu-celsius`.

**In summary**: in both v5.7.1 and v5.7.2, if I have "Use default" Regional preferences, the app works perfectly. After I change the Regional preferences away from default, the app will only crash. Changing Regional preferences back to "Use default" then results in the app again behaving normally.

@rdoeffinger I hope that's helpful for you, and hopefully also allows you to be able to remove the full ICU package and decrease the app size again (as in test3 and v5.7.2).

**Comment by code-consensus** [comment]
> The only thing I did notice with this test APK is that the app had some responsiveness issue -- though this may be because the file is not a release variant.

FYI I discovered that the responsiveness issue with selecting the dictionary flag [I mentioned above](https://github.com/rdoeffinger/Dictionary/issues/159#issuecomment-2326225569) is _not_ related to the changes in test2, test3 or v5.7.2 -- because I discovered this is _also_ the case in the older v5.7.1.

What is _nice_ about changing the dictionary by selecting the flag instead of three-dots/dictionary manager, is that it keeps the current word in the search field, so that you can instantly have the new dictionary you select search that term. 

But for whatever reason I'm getting a responsiveness issue on this, it is clearly not related to the dictionary crashing.

**Comment by code-consensus** [comment]
Hi @rdoeffinger , I can confirm that [my workaround posted above](https://github.com/rdoeffinger/Dictionary/issues/159#issuecomment-2355152344) (using the smaller codebase of v5.7.2/5.7.1) also works under **Android v15** as currently released in AP3A.241005.015.

**Comment by code-consensus** [comment]
Hi @rdoeffinger, not sure if wanting to revert back away from the full size ICU code, but for my workaround that I [posted above](https://github.com/rdoeffinger/Dictionary/issues/159#issuecomment-2355152344) (the source of the crash being caused by the use of regional settings which alters the locale with a "#"), I can confirm that the workaround also now works with Android v16 as deployed. I tested the older versions with the smaller code base (e.g. v5.7.2). 

Testing a number of locales and languages, v5.7.2 would not crash -- as soon as I would add a regional preferences (e.g. specify temperature unit, change first day of the week, or now with Android v16 you can specify metric system), the program would do the infinite loop crashing. As soon as I change it all back to default, app doesn't crash.

Just FYI if you wanted to go back to the smaller code base in case a fix could be made!
